### PR TITLE
Subsurface scattering

### PR DESCRIPTION
**Pull request by CraigFeldspar**
+ Adds a subsurface scattering post process
+ Enables to render the scene directly on a multiple render target, with additionnal RT being used to store normals, depth, light info etc..
+ Fixes a bug with MSAA on multiRenderTarget textures

### CODE REVIEW

**Comment by sebavan** [comment]
Could you reopen the PR as a draft ? might prevent early merge ?

**Comment by sebavan** [comment]
Also we probably should rename HIGH_DEFINITION_PIPELINE to DEFERED_PIPELINE to be future looking and allow it to default to half float vs float .

From the shader, I guess this would not work in Webgl1 which I do not mind but might be a good idea to emit a warning/error in this case ?

I ll look more in detail once reopened as a draft but the big picture looks overall good :-)

**Comment by sebavan** [comment]
Review has been done offline with more ideas to prevent changing all the materials.

**Comment by CraigFeldspar** [comment]
# 2 PG are available for testing :  

## Neutral diffusion profile on stanford dragon (kinda jade material) : 
http://localhost:1338/Playground/index-local.html#GTQKYK

## Skin diffusion profile on Lee Perry Smith head
http://localhost:1338/Playground/index-local.html#W7DYG2

I'm still not removing draft status because there is still some minor bugs remaining : 
+ Material becomes blue when metallic is set to 1 and SSS is enabled
+ There is a ghost frame when activating SSS, while the shader is compiling
+ WebGL 1 needs to be disabled gracefully

@sebavan we're in touch about that offline tomorrow :)

File: src/Rendering/prePassRenderer.ts [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+        // PDF[r, phi, s] = r * R[r, phi, s]
+        // CDF[r, s]      = 1 - 1/4 * Exp[-r * s] - 3/4 * Exp[-r * s / 3]
+        // ------------------------------------------------------------------------------------
+        // We importance sample the color channel with the widest scattering distance.
+        const maxScatteringDistance = Math.max(color.r, color.g, color.b);
+
+        return this._sampleBurleyDiffusionProfile(cdf, maxScatteringDistance);
+    }
+
+    // https://zero-radiance.github.io/post/sampling-diffusion/
```
Comment (sebavan) [inline src/Rendering/prePassRenderer.ts:344]: move one function up for the refer

File: src/Rendering/prePassRendererSceneComponent.ts [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+Scene.prototype.enablePrePassRenderer = function(): Nullable<PrePassRenderer> {
+    if (this._prePassRenderer) {
+        return this._prePassRenderer;
+    }
+
+    this._prePassRenderer = new PrePassRenderer(this);
+
+    if (!this._prePassRenderer.isSupported) {
+        this._prePassRenderer = null;
+        Logger.Error("Pre-pass needs WebGL 2.");
```
Comment (sebavan) [inline src/Rendering/prePassRendererSceneComponent.ts:62]: Add pre pass in the sentence and group in one error

File: src/Shaders/ShadersInclude/prePassDeclaration.fx [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#ifdef PREPASS
+#if __VERSION__ >= 200
```
Comment (sebavan) [inline src/Shaders/ShadersInclude/prePassDeclaration.fx:2]: remove as PREPASS only available in 2.0

File: src/Shaders/pbr.fragment.fx [hunk 4]
```diff
@@ -492,12 +495,32 @@ void main(void) {
...
+    gl_FragData[1] = vec4(tagLightingForSSS(irradiance), scatteringDiffusionProfile / 255.); // Irradiance + SS diffusion profile
+    #else
+    gl_FragData[1] = vec4(irradiance, 1.0); // Irradiance
+    #endif
+    gl_FragData[2] = vec4(vViewPos.z, 0.0, 0.0, 1.0); // Linear depth
+    gl_FragData[3] = vec4(sqAlbedo, 1.0); // albedo, for pre and post scatter
+#endif
+
+    gl_FragColor = finalColor;
     #include<pbrDebug>
```
Comment (sebavan) [inline src/Shaders/pbr.fragment.fx:525]: Debug to troubleshoot.

File: src/Shaders/sceneCompositor.fragment.fx [hunk 5]
```diff
@@ -0,0 +1,8 @@
+﻿// Samplers
```
Comment (sebavan) [inline src/Shaders/sceneCompositor.fragment.fx:1]: To Remove

File: src/Shaders/subSurfaceScattering.fragment.fx [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    for (int i = 0; i < n; i++)
+    {
+        // Integrate over the image or tangent plane in the view space.
+        EvaluateSample(i, n, S, d, centerPosVS, mmPerUnit, pixelsPerMm,
+                       phase, totalIrradiance, totalWeight);
+    }
+    // Total weight is 0 for color channels without scattering.
+    totalWeight = max(totalWeight, HALF_MIN);
+
+    // gl_FragColor = vec4(totalIrradiance / totalWeight, 1.);
```
Comment (sebavan) [inline src/Shaders/subSurfaceScattering.fragment.fx:232]: Comments to remove

**Comment by sebavan** [comment]
Comments might seem short but did the review in pair programming with @CraigFeldspar  :-) and this is going to be great !!!!

File: src/Engines/Extensions/engine.multiRender.ts [hunk 7]
```diff
@@ -28,15 +30,54 @@ declare module "../../Engines/thinEngine" {
...
-        updateMultipleRenderTargetTextureSampleCount(textures: Nullable<InternalTexture[]>, samples: number): number;
+        updateMultipleRenderTargetTextureSampleCount(count: number, textures: Nullable<InternalTexture[]>, samples: number): number;
+
+        /**
+         * Clears attachments
+         * @param texture texture to clear
+         * @param attachments attachments to clear
+         * @param color Clear color
+         */
+        clearColorAttachments(texture: InternalTexture, attachments: number[], color?: IColor4Like) : void;
```
[the remark this answers is not in the document]
Comment (CraigFeldspar) [reply src/Engines/Extensions/engine.multiRender.ts]: function removed

File: src/Engines/Extensions/engine.multiRender.ts [hunk 8]
```diff
@@ -28,15 +30,54 @@ declare module "../../Engines/thinEngine" {
...
 }
 
-ThinEngine.prototype.unBindMultiColorAttachmentFramebuffer = function(textures: InternalTexture[], disableGenerateMipMaps: boolean = false, onBeforeUnbind?: () => void): void {
+ThinEngine.prototype.renderToAttachments = function(attachments: any[]): void {
+    const gl = this._gl;
+
+    gl.drawBuffers(attachments);
+};
+
+ThinEngine.prototype.clearColorAttachments = function(texture: InternalTexture, attachments: number[], color?: IColor4Like): void {
```
[the remark this answers is not in the document]
Comment (CraigFeldspar) [reply src/Engines/Extensions/engine.multiRender.ts]: function removed

File: src/Materials/effect.ts [hunk 9]
```diff
@@ -139,6 +143,8 @@ export class Effect implements IDisposable {
 
     /** @hidden */
     public _bonesComputationForcedToCPU = false;
+    /** @hidden */
+    public _multiTarget: boolean = false;
```
[the remark this answers is not in the document]
Comment (CraigFeldspar) [reply src/Materials/effect.ts:147]: Unfortunately this is the easiest place to store it if we don't want to add dirty checks on shader hot-swap status on materials :(

File: src/Engines/Extensions/engine.multiRender.ts [hunk 10]
```diff
@@ -73,7 +82,7 @@ ThinEngine.prototype.unBindMultiColorAttachmentFramebuffer = function(textures:
         gl.drawBuffers(attachments);
     }
 
-    for (var i = 0; i < textures.length; i++) {
+    for (var i = 0; i < textures[0]._attachments!.length; i++) {
```
Comment (sebavan) [inline src/Engines/Extensions/engine.multiRender.ts:85]: [NIT] extract `const count = textures[0]._attachments!.length;` at the top so you could use count here as well.

File: src/PostProcesses/subSurfaceScatteringPostProcess.ts [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    /** @hidden */
+    public texelWidth: number;
+    /** @hidden */
+    public texelHeight: number;
+
+    constructor(name: string, scene: Scene, options: number | PostProcessOptions, camera: Nullable<Camera> = null, samplingMode?: number, engine?: Engine, reusable?: boolean, textureType: number = Constants.TEXTURETYPE_UNSIGNED_INT) {
+        super(name, "subSurfaceScattering", ["texelSize", "viewportSize", "metersPerUnit"], ["diffusionS", "diffusionD", "filterRadii", "irradianceSampler", "depthSampler", "albedoSampler"], options, camera, samplingMode || Texture.BILINEAR_SAMPLINGMODE, engine, reusable, null, textureType, "postprocess", undefined, true);
+        this._scene = scene;
+
+        const defines = this._getDefines();
```
Comment (sebavan) [inline src/PostProcesses/subSurfaceScatteringPostProcess.ts:28]: I guess we could just avoid with dealing with defines which are always empty ?

File: src/PostProcesses/subSurfaceScatteringPostProcess.ts [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+                Math.tan(scene.activeCamera!.fov / 2) * scene.getEngine().getAspectRatio(scene.activeCamera!, true),
+                Math.tan(scene.activeCamera!.fov / 2));
+            effect.setArray3("diffusionS", scene.prePassRenderer.ssDiffusionS);
+            effect.setArray("diffusionD", scene.prePassRenderer.ssDiffusionD);
+            effect.setArray("filterRadii", scene.prePassRenderer.ssFilterRadii);
+        });
+
+    }
+
+    private _getDefines(): Nullable<string> {
```
Comment (sebavan) [inline src/PostProcesses/subSurfaceScatteringPostProcess.ts:52]: I guess we could just avoid with dealing with defines which are always empty ?

File: src/Shaders/ShadersInclude/fibonacci.fx [hunk 13]
```diff
@@ -0,0 +1,16 @@
+#define rcp(x) 1. / x
+#define GOLDEN_RATIO 1.618033988749895
+#define TWO_PI 6.2831855
+
```
Comment (sebavan) [inline src/Shaders/ShadersInclude/fibonacci.fx:4]: Add a quick link to the source pdf ? as a helper in case of changes ?

**Comment by sebavan** [comment]
Actually, last point, is it possible to split the prepass renderer and the subsurface specific infos ???

It feels weird to set the metersPerUnit on the prepass which could be generic for deffered ?

**Comment by CraigFeldspar** [comment]
> Actually, last point, is it possible to split the prepass renderer and the subsurface specific infos ???
> 
> It feels weird to set the metersPerUnit on the prepass which could be generic for defered ?
> 
> It might not be so no worry we ll check if we could make it more generic later on.

You're totally right, as this will be the renderer for several effect.

What about prePassRender.xxxConfiguration ?

So we could have subSurfaceConfiguration, ssaoConfiguration etc...

**Comment by sebavan** [comment]
I love it, please GO GO GO !!!!

File: src/scene.ts [hunk 14]
```diff
@@ -884,6 +884,11 @@ export class Scene extends AbstractScene implements IAnimatable {
     */
     public fogEnd = 1000.0;
 
+    /**
+    * Flag indicating that the frame buffer binding is handled by another component
+    */
+    public prePass : boolean = false;
```
Comment (deltakosh) [inline src/scene.ts:890]: nit: 
```suggestion
    public prePass: boolean = false;

```

File: src/Engines/Extensions/engine.multiRender.ts [hunk 15]
```diff
@@ -67,13 +76,15 @@ ThinEngine.prototype.unBindMultiColorAttachmentFramebuffer = function(textures:
...
+
+        for (var i = 0; i < count; i++) {
             attachments[i] = (<any>gl)[this.webGLVersion > 1 ? "COLOR_ATTACHMENT" + i : "COLOR_ATTACHMENT" + i + "_WEBGL"];
         }
+
         gl.drawBuffers(attachments);
     }
 
-    for (var i = 0; i < textures.length; i++) {
+    for (var i = 0; i < textures[0]._attachments!.length; i++) {
```
Comment (sebavan) [inline src/Engines/Extensions/engine.multiRender.ts:87]: This should be count as you have it accessible now :-)

File: src/Rendering/prePassRenderer.ts [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+    private _defaultAttachments: number[];
+    private _clearAttachments: number[];
+
+    private readonly _clearColor = new Color4(0, 0, 0, 0);
+
+    /**
+     * Defines the ratio real world => scene units.
+     * Used for subsurface scattering
+     */
+    public metersPerUnit: number = 1;
```
Comment (sebavan) [inline src/Rendering/prePassRenderer.ts:54]: this should be in the subsurface configuration ?

File: src/Rendering/prePassRendererSceneComponent.ts [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+         */
+        disablePrePassRenderer(): void;
+
+        /**
+         * Diffusion profile colors for subsurface scattering
+         * You can add one diffusion color using `addDiffusionProfile` on `scene.prePassRenderer`
+         * See ...
+         * Note that you can only store up to 5 of them
+         */
+        ssDiffusionProfileColors: Color3[];
```
Comment (sebavan) [inline src/Rendering/prePassRendererSceneComponent.ts:36]: Shouldnt this be part of the subsurfaceConfiguration in the PrePassRenderer ?

Comment (CraigFeldspar) [reply src/Rendering/prePassRendererSceneComponent.ts:36]: I put it in the scene in order for that variable to be serialized. It reuses the pattern from `AudioSceneComponent`.
Maybe we should look into how to serialize infos in the component itself ? I don't think it has been done yet

File: src/Rendering/prePassRenderer.ts (hunk 16 above)
Comment (CraigFeldspar) [reply src/Rendering/prePassRenderer.ts:54]: it should :)

File: src/Rendering/prePassRendererSceneComponent.ts (hunk 17 above)
Comment (CraigFeldspar) [reply src/Rendering/prePassRendererSceneComponent.ts]: nevermind i found a way !

**Comment by CraigFeldspar** [comment]
I also fixed a bug on postprocesses that caused the playground to infinite loop on reload

**Comment by sebavan** [comment]
@CraigFeldspar as soon as the build is green we can merge it !!!

**Comment by sebavan** [comment]
Helll Yeah !!!!!

**Pull request merged**